### DISCUSSION TITLE
Enable nesting by default

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12130] [Ideas]
Dagger natively supports nesting (dagger in dagger). At the moment this is an opt-in feature (`Container.withExec(experimentalPrivilegedNesting: true)`. This is as a precaution while we finalized the feature and made sure it's properly sandboxed.

Are we now able to say "mission accomplished" and turn it on by default? If not, what's the path to getting there?

**Comment by marcosnils** [comment]
SGTM I don't see any reasons why "not" doing this.

**Comment by sipsma** [comment]
Right now when an exec has nesting enabled the cache key for it includes the engine version, so it's cache invalidated when the engine version changes (i.e. an upgrade or a rebuild w/ changes when doing local dev on the engine).

IIRC the main motivation for having this caching behavior was our own integ tests (which make very heavy use of nested execs) and wanting the tests to be invalidated when we modified the code in the engine API. However, we could accomplish that for our own tests in a different way.

Now that I think of it, I can't really come up with a compelling reason an end user (outside our integ tests) would need this behavior. The best I can think of would be if the behavior of an API changed and the user wanted to force a re-run of their nested exec using the new behavior without any changes to their actual code. That's pretty obscure though and also has other escape hatches (just clear the cache).

So provided we:
1. get rid of incorporating the engine version into the exec cache key when nesting is enabled
2. update our integ tests to deal with invalidating after engine changes

I think I'd be good with enabling it by default.

---

Another *possible* gotcha to look out for is any performance degradation since there's a little more overhead to nested execs than non-nested ones, but I would be surprised if it was a meaningful bottleneck. Just something to keep an eye on if we do this.

**Comment by shykes** [comment]
Outstanding! Since this is well-scoped and doesn't require any design bikeshedding, I'm marking it as a good first contribution.
